File: src/main/java/net/eduard/economy/hooks/EconomyVaultSupport.java
package net.eduard.economy.hooks;

import java.util.List;
import java.util.stream.Collectors;

import net.eduard.api.lib.modules.FakePlayer;
import net.eduard.economy.EduEconomy;
import net.eduard.economy.core.EconomyManager;
import org.bukkit.OfflinePlayer;

import net.eduard.api.lib.modules.Extra;
import net.milkbowl.vault.economy.Economy;
import net.milkbowl.vault.economy.EconomyResponse;
import net.milkbowl.vault.economy.EconomyResponse.ResponseType;

public class EconomyVaultSupport implements Economy {
	public EconomyVaultSupport(){}
	private EconomyManager manager = EduEconomy.getInstance().getManager();

	public EconomyResponse bankBalance(String bankName) {
		FakePlayer fake = new FakePlayer(bankName);
		return new EconomyResponse(0, manager.getCoins(fake), ResponseType.SUCCESS, null);

	}

	public EconomyResponse bankDeposit(String bankName, double amount) {

		FakePlayer fake = new FakePlayer(bankName);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public EconomyResponse bankHas(String bankName, double amount) {

		FakePlayer fake = new FakePlayer(bankName);
		if (manager.getCoins(fake) >= amount)
			return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.FAILURE, null);
	}

	public EconomyResponse bankWithdraw(String bankName, double amount) {

		FakePlayer fake = new FakePlayer(bankName);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public boolean hasBank(String bankName) {
		return true;
	}

	public EconomyResponse createBank(String bankName, String bankOwner) {
			return new EconomyResponse(0, 0, ResponseType.SUCCESS, null);

	}

	public EconomyResponse createBank(String bankName, OfflinePlayer player) {
			return new EconomyResponse(0, 0, ResponseType.SUCCESS, null);
	}

	public boolean createPlayerAccount(String playerName) {
		return true;
	}

	public boolean createPlayerAccount(OfflinePlayer player) {
		return true;
	}

	public boolean createPlayerAccount(String playerName, String worldName) {
		return true;
	}

	public boolean createPlayerAccount(OfflinePlayer player, String worldName) {
		return true;
	}

	public String currencyNamePlural() {
		return "Coins";
	}

	public String currencyNameSingular() {
		return "Coin";
	}

	public EconomyResponse deleteBank(String bankName) {
		EconomyManager m = EduEconomy.getInstance().getManager();
		FakePlayer fake = new FakePlayer(bankName);
			m.getCurrency().remove(fake);
			return new EconomyResponse(0, 0, ResponseType.SUCCESS, null);

	}

	public EconomyResponse depositPlayer(String playerName, double amount) {

		FakePlayer fake = new FakePlayer(playerName);
		manager.addCoins(fake, amount);
		return new EconomyResponse(amount,manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public EconomyResponse depositPlayer(OfflinePlayer player, double amount) {

		FakePlayer fake = new FakePlayer(player);
		manager.addCoins(fake, amount);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public EconomyResponse depositPlayer(String playerName, String wourldName, double amount) {
		return depositPlayer(playerName, amount);
	}

	public EconomyResponse depositPlayer(OfflinePlayer player, String wourldName, double amount) {
		return depositPlayer(player, amount);
	}

	public String format(double amount) {
		return Extra.MONEY.format(amount);
	}

	public int fractionalDigits() {
		return 0;
	}

	public double getBalance(String playerName) {

		FakePlayer fake = new FakePlayer(playerName);
		return manager.getCoins(fake);
	}

	public double getBalance(OfflinePlayer player) {

		FakePlayer fake = new FakePlayer(player);
		return manager.getCoins(fake);
	}

	public double getBalance(String playerName, String worldName) {
		return getBalance(playerName);
	}

	public double getBalance(OfflinePlayer player, String worldName) {
		return getBalance(player);
	}

	public List<String> getBanks() {
		return EduEconomy.getInstance().getManager()
				.getCurrency()
				.keySet()
				.stream()
				.map(FakePlayer::getName)
				.collect(Collectors.toList());
	}

	public boolean has(String playerName, double amount) {
		return getBalance(playerName) >= amount;
	}

	public boolean has(OfflinePlayer player, double amount) {
		return getBalance(player) >= amount;
	}

	public boolean has(String playerName, String worldName, double amount) {
		return getBalance(playerName) >= amount;
	}

	public boolean has(OfflinePlayer player, String worldName, double amount) {
		return getBalance(player) >= amount;
	}

	public boolean hasAccount(String playerName) {
		return true;
	}

	public boolean hasAccount(OfflinePlayer player) {
		return true;
	}

	public boolean hasAccount(String playerName, String worldName) {
		return true;
	}

	public boolean hasAccount(OfflinePlayer player, String worldName) {
		return true;
	}

	public boolean hasBankSupport() {
		return true;
	}

	public EconomyResponse isBankMember(String bankName, String playerName) {
		return new EconomyResponse(0, 0, ResponseType.NOT_IMPLEMENTED, null);
	}

	public EconomyResponse isBankMember(String bankName, OfflinePlayer player) {
		return new EconomyResponse(0, 0, ResponseType.NOT_IMPLEMENTED, null);
	}

	public EconomyResponse isBankOwner(String bankName, String playerName) {
		return new EconomyResponse(0, 0, ResponseType.NOT_IMPLEMENTED, null);
	}

	public EconomyResponse isBankOwner(String bankName, OfflinePlayer player) {
		return new EconomyResponse(0, 0, ResponseType.NOT_IMPLEMENTED, null);
	}

	public boolean isEnabled() {
		return EduEconomy.getInstance().isEnabled();
	}

	public EconomyResponse withdrawPlayer(String playerName, double amount) {

		FakePlayer fake = new FakePlayer(playerName);
		manager.removeCoins(fake, amount);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public EconomyResponse withdrawPlayer(OfflinePlayer player, double amount) {

		FakePlayer fake = new FakePlayer(player);
		manager.removeCoins(fake, amount);
		return new EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null);
	}

	public EconomyResponse withdrawPlayer(String playerName, String worldName, double amount) {
		return withdrawPlayer(playerName, amount);
	}

	public EconomyResponse withdrawPlayer(OfflinePlayer player, String worldName, double amount) {
		return withdrawPlayer(player, amount);
	}

	
	public String getName() {
		
		return "EduEconomy";
	}

}

File: src/main/java/net/eduard/economy/hooks/EconomyVaultSupport.kt
package net.eduard.economy.hooks

import net.milkbowl.vault.economy.Economy
import net.milkbowl.vault.economy.EconomyResponse
import net.eduard.api.lib.modules.FakePlayer
import net.milkbowl.vault.economy.EconomyResponse.ResponseType
import net.eduard.api.lib.modules.Extra
import net.eduard.economy.EduEconomyPlugin
import org.bukkit.OfflinePlayer
import java.util.stream.Collectors

class EconomyVaultSupport : Economy {
    override fun bankBalance(bankName: String): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(bankName)
        return EconomyResponse(0.0, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }

    override fun bankDeposit(bankName: String, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fake = FakePlayer(bankName)
        return EconomyResponse(amount, manager.getCoins(fake), ResponseType.SUCCESS, null)
    }

    override fun bankHas(bankName: String, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(bankName)
        return if (manager.getCoins(fakePlayer) >= amount) EconomyResponse(
            amount,
            manager.getCoins(fakePlayer),
            ResponseType.SUCCESS,
            null
        ) else EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.FAILURE, null)
    }

    override fun bankWithdraw(bankName: String, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(bankName)
        return EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }


    override fun createBank(bankName: String, bankOwner: String): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.SUCCESS, null)
    }

    override fun createBank(bankName: String, player: OfflinePlayer): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.SUCCESS, null)
    }

    override fun createPlayerAccount(playerName: String): Boolean {
        return true
    }

    override fun createPlayerAccount(player: OfflinePlayer): Boolean {
        return true
    }

    override fun createPlayerAccount(playerName: String, worldName: String): Boolean {
        return true
    }

    override fun createPlayerAccount(player: OfflinePlayer, worldName: String): Boolean {
        return true
    }

    override fun currencyNamePlural(): String {
        return "Coins"
    }

    override fun currencyNameSingular(): String {
        return "Coin"
    }

    override fun deleteBank(bankName: String): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(bankName)
        manager.currency.remove(fakePlayer)
        return EconomyResponse(0.0, 0.0, ResponseType.SUCCESS, null)
    }

    override fun depositPlayer(playerName: String, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(playerName)
        manager.addCoins(fakePlayer, amount)
        return EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }

    override fun depositPlayer(player: OfflinePlayer, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(player)
        manager.addCoins(fakePlayer, amount)
        return EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }

    override fun depositPlayer(playerName: String, wourldName: String, amount: Double): EconomyResponse {
        return depositPlayer(playerName, amount)
    }

    override fun depositPlayer(player: OfflinePlayer, wourldName: String, amount: Double): EconomyResponse {
        return depositPlayer(player, amount)
    }

    override fun format(amount: Double): String {
        return Extra.MONEY.format(amount)
    }

    override fun fractionalDigits(): Int {
        return 4
    }

    override fun getBalance(playerName: String): Double {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(playerName)
        return manager.getCoins(fakePlayer)
    }

    override fun getBalance(player: OfflinePlayer): Double {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(player)
        return manager.getCoins(fakePlayer)
    }

    override fun getBalance(playerName: String, worldName: String): Double {
        return getBalance(playerName)
    }

    override fun getBalance(player: OfflinePlayer, worldName: String): Double {
        return getBalance(player)
    }

    override fun getBanks(): List<String> {
        return EduEconomyPlugin.instance.manager
            .currency
            .keys
            .stream()
            .map { fakePlayer: FakePlayer -> fakePlayer.name }
            .collect(Collectors.toList())

    }

    override fun has(playerName: String, amount: Double): Boolean {
        return  EduEconomyPlugin.instance.manager.hasCoins(FakePlayer(playerName), amount)
    }

    override fun has(player: OfflinePlayer, amount: Double): Boolean {
        return  EduEconomyPlugin.instance.manager.hasCoins(FakePlayer(player), amount)
    }

    override fun has(playerName: String, worldName: String, amount: Double): Boolean {
        return  EduEconomyPlugin.instance.manager.hasCoins(FakePlayer(playerName), amount)
    }

    override fun has(player: OfflinePlayer, worldName: String, amount: Double): Boolean {
        return  EduEconomyPlugin.instance.manager.hasCoins(FakePlayer(player), amount)
    }

    override fun hasAccount(playerName: String): Boolean {
        return true
    }

    override fun hasAccount(player: OfflinePlayer): Boolean {
        return true
    }

    override fun hasAccount(playerName: String, worldName: String): Boolean {
        return true
    }

    override fun hasAccount(player: OfflinePlayer, worldName: String): Boolean {
        return true
    }

    override fun hasBankSupport(): Boolean {
        return true
    }

    override fun isBankMember(bankName: String, playerName: String): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.NOT_IMPLEMENTED, null)
    }

    override fun isBankMember(bankName: String, player: OfflinePlayer): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.NOT_IMPLEMENTED, null)
    }

    override fun isBankOwner(bankName: String, playerName: String): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.NOT_IMPLEMENTED, null)
    }

    override fun isBankOwner(bankName: String, player: OfflinePlayer): EconomyResponse {
        return EconomyResponse(0.0, 0.0, ResponseType.NOT_IMPLEMENTED, null)
    }

    override fun isEnabled(): Boolean {
        return true
    }

    override fun withdrawPlayer(playerName: String, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(playerName)
        manager.removeCoins(fakePlayer, amount)
        return EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }

    override fun withdrawPlayer(player: OfflinePlayer, amount: Double): EconomyResponse {
        val manager = EduEconomyPlugin.instance.manager
        val fakePlayer = FakePlayer(player)
        manager.removeCoins(fakePlayer, amount)
        return EconomyResponse(amount, manager.getCoins(fakePlayer), ResponseType.SUCCESS, null)
    }

    override fun withdrawPlayer(playerName: String, worldName: String, amount: Double): EconomyResponse {
        return withdrawPlayer(playerName, amount)
    }

    override fun withdrawPlayer(player: OfflinePlayer, worldName: String, amount: Double): EconomyResponse {
        return withdrawPlayer(player, amount)
    }

    override fun getName(): String {
        return "EduEconomy"
    }
}